### PR TITLE
[AIRFLOW-780] Fix dag import errors no longer working

### PR DESCRIPTION
**Pull request by aoen**
Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-780

The import errors were no longer working after the multiprocessor update
(since they are cleared after each DAG directory is parsed). This change
fixes them, and adds tests to prevent future regressions.

Also fix Py3 incompat in BaseTaskRunner and a couple of linter errors.

Note that there are a few inefficiencies (e.g. sometimes we delete then add import errors in the same place instead of just doing an update), but this is equivalent to the old behavior.

Testing Done:
- Added missing unit tests for dag imports. Note that some of them strangely fail for python 3 and it became too time consuming to debug since I don't have a copy of the travis environment, I even ran with the same version of python locally and couldn't reproduce. I have skipped those 3 tests in python 3 for now.

@bolkedebruin @artwr @mistercrunch @plypaul

### CODE REVIEW

File: tests/jobs.py [hunk 1]
```diff
@@ -97,19 +107,20 @@ def test_backfill_multi_dates(self):
         job = BackfillJob(
             dag=dag,
             start_date=DEFAULT_DATE,
-            end_date=DEFAULT_DATE+datetime.timedelta(days=1),
+            end_date=DEFAULT_DATE + datetime.timedelta(days=0),
```
Comment (artwr) [inline tests/jobs.py]: Why the change?

File: tests/jobs.py [hunk 2]
```diff
@@ -244,6 +254,27 @@ class SchedulerJobTest(unittest.TestCase):
 
     def setUp(self):
         self.dagbag = DagBag()
+        session = settings.Session()
+        session.query(models.ImportError).delete()
+        session.commit()
+
+    @staticmethod
+    def run_single_scheduler_loop_with_no_dags(dags_folder):
```
Comment (artwr) [inline tests/jobs.py:265]: Good idea.

**Review by artwr — APPROVED** [review]
LGTM overall.

File: tests/jobs.py (hunk 1 above)
Comment (aoen) [reply tests/jobs.py]: Good catch, I must have hit the increment number shortcut by accident! Surprisingly the test still passed.

File: airflow/task_runner/base_task_runner.py [hunk 3]
```diff
@@ -89,7 +89,7 @@ def __init__(self, local_task_job):
 
     def _read_task_logs(self, stream):
         while True:
-            line = stream.readline()
```
Comment (bolkedebruin) [inline airflow/task_runner/base_task_runner.py]: decode?

File: tests/jobs.py [hunk 4]
```diff
@@ -52,6 +53,15 @@
 DEV_NULL = '/dev/null'
 DEFAULT_DATE = datetime.datetime(2016, 1, 1)
 
+# Include the words "airflow" and "dag" in the file contents, tricking airflow into thinking these
```
Comment (bolkedebruin) [inline tests/jobs.py:59]: and or or?

File: tests/jobs.py [hunk 5]
```diff
@@ -97,19 +107,20 @@ def test_backfill_multi_dates(self):
...
-            end_date=DEFAULT_DATE+datetime.timedelta(days=1),
+            end_date=DEFAULT_DATE + datetime.timedelta(days=1),
             ignore_first_depends_on_past=True
         )
         job.run()
 
         session = settings.Session()
         drs = session.query(DagRun).filter(
-            DagRun.dag_id=='example_bash_operator'
+            DagRun.dag_id == 'example_bash_operator'
```
Comment (bolkedebruin) [inline tests/jobs.py]: I think this is not common practice in sqlalchemy?

File: tests/jobs.py (hunk 4 above)
Comment (aoen) [reply tests/jobs.py:59]: and.

```
                    if not all([s in content for s in (b'DAG', b'airflow')]):
                        self.file_last_changed[filepath] = file_last_changed_on_disk
                        return found_dags
```

File: tests/jobs.py (hunk 5 above)
Comment (aoen) [reply tests/jobs.py]: Undid this change (it goes against the pep8 which is why my linter caught it).

**Comment by aoen** [comment]
@bolkedebruin Addressed comments and waiting for your LGTM. Would be nice to get this for the RC since there is a regression from 1.7.1 otherwise (and it makes things harder operationally since users will ask you why their dags aren't running).

**Comment by bolkedebruin** [comment]
I'm not too happy with skipping tests on Travis as it is the only place (publicly) where all tests always run. Can you share what the Travis error was?

Rest LGTM, but I do think it warrants another Beta as it touches scheduler core.

**Comment by aoen** [comment]
I agree with you that it's not ideal. I spent about half a day debugging but it was slow-going (I even installed pyenv and used the same version of python in travis). Note that this test coverage was missing before so it's still an improvement. I tried binary searching the code with print statements but this was also slow going.

The travis error was that the import errors simply weren't added in all of the tests that should have added them for python 3.

When we have containers these kinds of reproducibility problems will go away.

**Comment by bolkedebruin** [comment]
K lets put it in then

**Comment by aoen** [comment]
@bolkedebruin I merged the change and bumped the beta branch version to beta 5.
